File: Aoc_Kotlin/src/main/kotlin/day13/foldings/Day13.kt
package day13.foldings

import AoC_Lib.*
import java.lang.StringBuilder

object Day13: SomeDay(2021,13) {
    override val title = "Transparent Origami"

    data class Folding(val axis: Char, val i: Int)

    private fun foldAlign(sheet: Set<Pos>, fold: Folding): Set<Pos> =
        sheet.map { p -> when {
            fold.axis == 'x' && p.x > fold.i -> Pos(fold.i + fold.i - p.x, p.y)
            fold.axis == 'y' && p.y > fold.i -> Pos(p.x, fold.i + fold.i - p.y)
            else -> p } }.toSet()

    private fun sheetToString(sheet: Set<Pos>): String {
        val maxX = sheet.maxOf { it.x }
        val maxY = sheet.maxOf { it.y }
        val sb = StringBuilder()
            .append('\n')
        (0 .. maxY).forEach { y ->
            (0 .. maxX).forEach { x ->
                sb.append(if (Pos(x,y) in sheet) '#' else '.')}
            sb.append('\n') }
        return sb.toString()
    }

    override fun first(data: String): Any? {
        val (d,f) = data.split("\n\n")
        val dots = d.lines().map(Pos::fromString).toSet()
        val folds = f.lines().map { it.split('=')
            .let { Folding(it.first().last(), it.last().toInt()) } }
        return foldAlign(dots, folds.first()).size
    } // 602 Time: 46ms

    override fun second(data: String): Any? {
        val (d,f) = data.split("\n\n")
        val dots = d.lines().map(Pos::fromString).toSet()
        val folds = f.lines().map { it.split('=')
            .let { Folding(it.first().last(), it.last().toInt()) } }
        return sheetToString(folds.fold(dots,::foldAlign))
    }
    // .##...##..####...##.#..#.####..##..#..#
    // #..#.#..#.#.......#.#..#....#.#..#.#.#.
    // #....#..#.###.....#.####...#..#....##..
    // #....####.#.......#.#..#..#...#....#.#.
    // #..#.#..#.#....#..#.#..#.#....#..#.#.#.
    // .##..#..#.#.....##..#..#.####..##..#..#
    // Time: 6ms
}

fun main() = SomeDay.mainify(Day13)

File: AoC_Scala3/src/day13/day13.scala
package day13

import AoC_Lib.*
import scala.annotation.tailrec

type Dot = (Int,Int)
type Fold = (Char,Int)

object day13 {
  val Array(dots,folds) = inputStr(day = 13)
    .split("\n\n")
    //.splitByBlankLines()
  val sheet: Set[Dot] = dots
    .split('\n')
    .map{ case s"$x,$y" => x.toInt -> y.toInt }
    .toSet
  val foldings: Array[Fold] = folds
    .split('\n')
    .map{ case s"fold along $axis=$num" => axis.head -> num.toInt}

  def printSheet(sheet: Set[Dot]): String = {
    val pmax = sheet.max
    val sb = new StringBuilder()
    (0 to pmax._2).foreach{ y =>
        (0 to pmax._1).foreach{ x =>
            sb.append(if sheet.contains((x,y)) then '#' else ' ')}
        sb.append('\n')}
    sb.toString()
  }

  def foldAlong(sheet: Set[Dot], fold: Fold): Set[Dot] =
    sheet.map{ (x,y) => fold match {
      case ('x',v) if x > v => (v - (x - v), y)
      case ('y',v) if y > v => (x, v - (y - v))
      case _ => (x, y)
    }}

  def part1: Int = foldAlong(sheet,foldings.head).size

  def part2: String = printSheet(foldings.foldLeft(sheet)(foldAlong))

  @main
  def TransparentOrigami(): Unit = {
    Console.println(part1) // 602
    Console.println(part2) // 946346
  }
}
